tracks/elm/exercises/leap/mentoring.md

```diff
@@ -0,0 +1,90 @@
+This is a good exercise to discuss the describability of logic, as most students solve the problem using a conditional statement that is hard to describe in language, and there are other options.
+
+This exercise introduces `Basics.remainderBy`, the `let` / `in` syntax and partial application / currying.
+
+### Reasonable solutions
+
+The easiest / most common solution probably involves partially applying `Basics.remainderBy` and using a slightly complicated if statement:
+
+```elm
+isLeapYear : Int -> Bool
+isLeapYear year =
+    let
+        divisibleBy number = 
+            Basics.remainderBy number year == 0         
+    in
+        (
+            divisibleBy 4 && 
+            not (divisibleBy 100)
+        )
+        || divisibleBy 400
+```
+
+The complicated if statement is a bit of a problem, as it is hard to describe in language, and hence talking about the code with others is difficult.
+
+Using a case statement turns the if statement in to a truth table, and is probably a bit clearer, but at the expense of some extra code:
+
+```elm
+isLeapYear : Int -> Bool
+isLeapYear year =
+    let
+        divisibleBy number = 
+            Basics.remainderBy number year == 0         
+    in
+		case ( divisibleBy 4, divisibleBy 100, divisibleBy 400 ) of
+			( True, True, True ) ->
+				True
+
+			( True, True, False ) ->
+				False
+
+			( True, False, False ) ->
+				True
+
+			_ ->
+				False
+```
+
+If you think about the easiest way to describe the logic to somebody else, you probably come up with a solution like this:
+
+```elm
+isLeapYear : Int -> Bool
+isLeapYear year =
+    let
+        divisibleBy number = 
+            Basics.remainderBy number year == 0         
+    in
+		if divisibleBy 400 then
+			True
+
+		else if divisibleBy 100 then
+			False
+
+		else if divisibleBy 4 then
+			True
+
+		else
+			False
+```
+
+It is possible to go even further with the describability of the logic, using code like this:
+
+```elm
+isLeapYear : Int -> Bool
+isLeapYear year =
+	let
+		...
+	in
+		if is400YearSpecialCase then
+			True
+
+		else if is100YearSpecialCase then
+			False
+
+		else 
+			defaultCase
+```
+
+### Common suggestions
+
+Discussing the various options around the describability of the solution is worthwhile. The calculation of leap years is not going to change any time soon, so there is definitely an argument that the complicated conditional statement is ok. However most code deals with business logic, which changes much more frequently, and so the trade off's are different.
```

Comment:
I'm not sure if the first example, `divisibleBy 4 && (not (divisibleBy 100) || divisibleBy 400)`, is the complicated conditional statement. I don't think it is more complicated than the alternatives.

Reply:
Hi Simon `divisibleBy 4 && (not (divisibleBy 100) || divisibleBy 400)` is the complicated conditional statement I am talking about. I concede that it isn't very complicated, but I think it still illustrates the point. And if you try describing it to somebody using words I think you will find it difficult.